plonky2/src/hash/poseidon2_bn254.rs

```diff
@@ -1,149 +1,323 @@
+        if check_len_in_bytes(inputs) <= hash_size {
+            if inputs.len() == 1 {
+                // if there is one element and it is a BN field element return it.
+                if let GenericField::BN254(v) = inputs[0].clone() {
+                    return BN254HashOut{element: v};
+                }
+            }
+        }
+        // TODO: if we get 4 or less Goldilocks -> convert to BN return?
+        Self::hash_no_pad(inputs)
+    }
+
+
+    fn sponge(state: &mut Self::Permutation, input: Vec<GenericField<F>>) {
+
+        let bn_felts = generic_field_to_bn(&input);
+
+        // absorb in overwrite mode
+        for chunk in bn_felts.chunks(2) {
+            state.set_from_slice(chunk, 0);
+            state.permute();
+        }
+
     }
 
     fn two_to_one(left: Self::Hash, right: Self::Hash) -> Self::Hash {
-        let mut input_bytes = vec![0; N * 2];
-        input_bytes[0..N].copy_from_slice(&left.0);
-        input_bytes[N..].copy_from_slice(&right.0);
-        let mut arr = [0; N];
-        let state_u64: Vec<u64> = input_bytes
-            .chunks_exact(8)
-            .map(|chunk| u64::from_be_bytes(chunk.try_into().unwrap()))
-            .collect();
-
-        let hash = sponge_u64_no_pad(Hash::Poseidon2, state_u64);
-        let hash_bytes = felts_to_bytes(hash);
-        arr.copy_from_slice(&hash_bytes[..N]);
-        BytesHash(arr)
+
+        let mut perm = Self::Permutation::new(core::iter::repeat(BN254Fr::zero()));
+        perm.set_from_slice(&[left.element], 0);
+        perm.set_from_slice(&[right.element], 1);
+
+        perm.permute();
+        let out = perm.squeeze();
+
+        BN254HashOut {
+            element: out[0].clone(),
+        }
+    }
+
+    fn squeeze_goldilocks(state: &mut Self::Permutation) -> Vec<F> {
+        let bn_out = state.squeeze();
+        let bn_bytes: Vec<u8> = bn_out.iter().flat_map(|e| felts_to_bytes(e)).collect();
+        let goldilocks_felts: Vec<F> = bytes_to_u64(&bn_bytes).iter().map(|e| F::from_canonical_u64(*e)).collect();
+        assert!(goldilocks_felts.len()>0);
+        goldilocks_felts
+    }
+}
+
+// --------- Conversion helper functions ---------------------
+
+/// converts a vec of goldilocks to bn254
+/// takes 7 goldilocks and converts to 2 bn254
+fn goldilocks_to_bn<F: RichField>(input: &Vec<F>) -> Vec<BN254Fr>{
+    let u64s: Vec<u64> = input.iter().map(|x| x.to_canonical_u64()).collect();
+    let l = u64s.len();
+    let m = l / 7;
+    let mut result = Vec::new();
+
+    for i in 0..m {
+        let group: [u64; 7] = u64s[7 * i..7 * (i + 1)].try_into().unwrap();
+        let (a, b) = u64s_to_felts(group);
+        result.push(a);
+        result.push(b);
     }
+
+    let r = l - 7 * m;
+    if r > 0 {
+        let mut ws = [0u64; 7];
+        for i in 0..r {
+            ws[i] = u64s[7 * m + i];
+        }
+        let (a, b) = u64s_to_felts(ws);
+        // check that we don't push zero field elements
+        if a != BN254Fr::zero() {
```

Comment:
i don't think this conditional behaviour is a good idea. We need deterministic behaviour (especially as we will need to do the same in a circuit!)

- the simple solution: always push 2 BN elements
- the "more optimal" solution: check if (r <= 3), then push 1 BN, otherwise push 2 BN.